### ISSUE TITLE
No `totalPwrCsp` available

### ISSUE THREAD

**Issue by jwillemsen** [Hyundai-Kia-Connect/hyundai_kia_connect_api#440]
* Hyundai  Connect version: 2.0.16
* Hyundai Kona 2024 Full Electric

### Description

I see in the log that `totalPwrCsp` is not available. When using the code from your example 

```
vm.update_all_vehicles_with_cached_state()
print(vm.vehicles)
```

I get the following errror

```
Traceback (most recent call last):
  File "/home/johnny/ACE/github/hyundai_kia_connect_api/hyundai_kia_connect_api/KiaUvoApiEU.py", line 330, in update_vehicle_with_cached_state
    state = self._get_driving_info(token, vehicle)
  File "/home/johnny/ACE/github/hyundai_kia_connect_api/hyundai_kia_connect_api/KiaUvoApiEU.py", line 1059, in _get_driving_info
    drivingInfoItem["totalPwrCsp"]
KeyError: 'totalPwrCsp'
```

The data printed is the following (removed id and registration data)

```
{'d: Vehicle(id='d', name='KONA', model='KONA', registration_date='X', year=None, VIN='', key=None, enabled=True, _total_driving_range=None, _total_driving_range_value=None, _total_driving_range_unit=None, _odometer=None, _odometer_value=None, _odometer_unit=None, _geocode_address=None, _geocode_name=None, car_battery_percentage=100, engine_is_running=False, last_updated_at=datetime.datetime(2023, 11, 6, 14, 49, 15, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')), timezone=tzfile('/usr/share/zoneinfo/Europe/Berlin'), dtc_count=None, dtc_descriptions=None, smart_key_battery_warning_is_on=False, washer_fluid_warning_is_on=False, brake_fluid_warning_is_on=False, _air_temperature=21.0, _air_temperature_value=21.0, _air_temperature_unit='°C', air_control_is_on=False, defrost_is_on=False, steering_wheel_heater_is_on=False, back_window_heater_is_on=0, side_mirror_heater_is_on=None, front_left_seat_status='Off', front_right_seat_status='Off', rear_left_seat_status='Off', rear_right_seat_status='Off', is_locked=True, front_left_door_is_open=0, front_right_door_is_open=0, back_left_door_is_open=0, back_right_door_is_open=0, trunk_is_open=False, hood_is_open=False, front_left_window_is_open=0, front_right_window_is_open=0, back_left_window_is_open=0, back_right_window_is_open=0, tire_pressure_all_warning_is_on=False, tire_pressure_rear_left_warning_is_on=False, tire_pressure_front_left_warning_is_on=False, tire_pressure_front_right_warning_is_on=False, tire_pressure_rear_right_warning_is_on=False, _next_service_distance=None, _next_service_distance_value=None, _next_service_distance_unit=None, _last_service_distance=None, _last_service_distance_value=None, _last_service_distance_unit=None, _location_latitude=None, _location_longitude=None, _location_last_set_time=None, ev_charge_port_door_is_open=None, ev_charge_limits_dc=None, ev_charge_limits_ac=None, total_power_consumed=None, total_power_regenerated=None, power_consumption_30d=None, daily_stats=[], month_trip_info=None, day_trip_info=None, ev_battery_percentage=None, ev_battery_is_charging=None, ev_battery_is_plugged_in=None, _ev_driving_range=None, _ev_driving_range_value=None, _ev_driving_range_unit=None, _ev_estimated_current_charge_duration=None, _ev_estimated_current_charge_duration_value=None, _ev_estimated_current_charge_duration_unit='m', _ev_estimated_fast_charge_duration=None, _ev_estimated_fast_charge_duration_value=None, _ev_estimated_fast_charge_duration_unit='m', _ev_estimated_portable_charge_duration=None, _ev_estimated_portable_charge_duration_value=None, _ev_estimated_portable_charge_duration_unit='m', _ev_estimated_station_charge_duration=None, _ev_estimated_station_charge_duration_value=None, _ev_estimated_station_charge_duration_unit='m', _ev_target_range_charge_AC=None, _ev_target_range_charge_AC_value=None, _ev_target_range_charge_AC_unit=None, _ev_target_range_charge_DC=None, _ev_target_range_charge_DC_value=None, _ev_target_range_charge_DC_unit=None, ev_first_departure_enabled=None, ev_second_departure_enabled=None, ev_first_departure_days=None, ev_second_departure_days=None, ev_first_departure_time=None, ev_second_departure_time=None, ev_off_peak_start_time=None, ev_off_peak_end_time=None, ev_off_peak_charge_only_enabled=None, _fuel_driving_range=142, _fuel_driving_range_value=142, _fuel_driving_range_unit='km', fuel_level=0, fuel_level_is_low=False, engine_type=<ENGINE_TYPES.EV: 'EV'>, data={'vehicleStatus': {'windowRearCurtain': {}, 'transCond': True, 'airCtrlOn': False, 'engine': False, 'doorLock': True, 'doorOpen': {'frontLeft': 0, 'frontRight': 0, 'backLeft': 0, 'backRight': 0}, 'trunkOpen': False, 'airTemp': {'value': '0EH', 'unit': 0, 'hvacTempType': 1}, 'defrost': False, 'lowFuelLight': False, 'acc': False, 'ign3': True, 'hoodOpen': False, 'steerWheelHeatStep2': 0, 'steerWheelHeat': 0, 'sideBackWindowHeat': 0, 'dte': {'value': 142, 'unit': 1}, 'tirePressureLamp': {'tirePressureLampAll': 0, 'tirePressureLampFL': 0, 'tirePressureLampFR': 0, 'tirePressureLampRL': 0, 'tirePressureLampRR': 0}, 'seatHeaterVentState': {'frSeatHeatState': 2, 'flSeatHeatState': 2, 'rlSeatHeatState': 2, 'rrSeatHeatState': 2}, 'battery': {'batSoc': 100, 'bacState': 0, 'sjbDeliveryMode': 2}, 'lampWireStatus': {'stopLamp': {'leftLamp': False, 'rightLamp': False}, 'headLamp': {'headLampStatus': False, 'leftLowLamp': False, 'rigthLowLamp': False, 'leftHighLamp': False, 'rightHighLamp': False}, 'turnSignalLamp': {'leftFrontLamp': False, 'rightFrontLamp': False, 'leftRearLamp': False, 'rightRearLamp': False}}, 'windowOpen': {'frontLeft': 0, 'frontRight': 0, 'backLeft': 0, 'backRight': 0}, 'windowOpen2': {'frontLeft': 0, 'frontRight': 0, 'backLeft': 0, 'backRight': 0}, 'tailLampStatus': 0, 'hazardStatus': 0, 'smartKeyBatteryWarning': False, 'fuelLevel': 0, 'washerFluidStatus': False, 'breakOilStatus': False, 'sleepModeCheck': False, 'time': '20231106144915', 'airCleaning': {'airPurifierStatus': 0, 'fineDustStep': 0, 'fineDustColor': 0}}})}
```

**Comment by WillemD61** [comment]
What if you first run?
vm.force_refresh_all_vehicles_states()

If you can turn on logging you can see whether the field is available.
You can turn on logging by adding the following at line 52 of the EU api
logging.basicConfig(filename='EUapi.log', encoding='utf-8', level=logging.DEBUG)

The logging will be in the file EUapi.log. The totalPwrCsp should be in the response of get_driving_info responseAlltime

**Comment by jwillemsen** [comment]
Looks `totalPwrCsp` is not available, with logging enabled I do see the following for `get_driving_info`

```
DEBUG:urllib3.connectionpool:https://prd.eu-ccapi.hyundai.com:8080 "POST /api/v1/spa/vehicles/d5160cae-52f0-4acb-b256-568b206344e8/drvhistory HTTP/1.1" 200 169
DEBUG:hyundai_kia_connect_api.KiaUvoApiEU:hyundai_kia_connect_api - get_driving_info responseAlltime {'retCode': 'S', 'resCode': '0000', 'resMsg': {'drivingInfo': [{'drivingPeriod': 0}, {'drivingPeriod': 1}], 'drivingInfoDetail': []}, 'msgId': 'xx'}
```

**Comment by WillemD61** [comment]
Looks like you are not getting any data at all. No data for any driving period. No message id.

Did you activate your car at all in the App and are you using the correct authorisation data for the API? 
Is this the first day of driving? Maybe it takes time to come available? Maybe available tomorrow?

Sorry, I have run out of other suggestions.

**Comment by jwillemsen** [comment]
I cleared the msgId, not sure whether that could leak something. The bluelink app does work and I do see some other data in HA through this integration, but not all data, very likely Hyundai made some API changes, the bluelink app isn't working as good with the new 2024 Kona as it did with the i10

**Comment by cdnninja** [comment]
Before that line does it contain other car data? If so I think we just need to adjust code to handle no driving data.

**Comment by jwillemsen** [comment]
Yes, other calls do return data, in the bluelink app I see more data compared to HA

**Comment by WillemD61** [comment]
Another thought: Did you authorise (give consent) data sharing in the app?

**Comment by jwillemsen** [comment]
Yes, I did that, I can see the status of the car in HA (doors/windows/etc)

**Comment by cdnninja** [comment]
I assume your car supports driving info.  I assume the native app isn't showing driving or usage data?

**Comment by jwillemsen** [comment]
The bluelink app does show driving info and power usage, but not when I use this integration in home assistant

**Comment by WillemD61** [comment]
Can you show, for comparison, the full contents of your log file (with any sensitive info crossed out). The full cycle of data collection should show 3 times "Starting new HTTPS connection" followed by a GET or POST command and then the response.

**Comment by jwillemsen** [comment]
Attached the log, I did run force_refresh_all_vehicles_states

[EUapi.log](https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api/files/13464221/EUapi.log)

**Comment by WillemD61** [comment]
I compared your log to mine. It is clear this is not just about totalPwrCsp. There is so much info missing in the responses you are getting thorugh the API, but also you have a few fields that I don't have.
I have no idea why this is and what I find strange is that your are getting the info in the APP but not in the API, so the info exists but you are not getting it through the API.  Maybe new/different API requests are needed?
Further analysis is beyond my knowledge, sorry.

**Comment by jwillemsen** [comment]
Very likely a API change because this is a 2024 Kona model

**Comment by fuatakgun** [comment]
There were some historical discussions between api v1 and v2 versions, we decided to stick with v1 as it was functional enough :-) apparently not enough

**Comment by jwillemsen** [comment]
Is there some documentation to read, maybe I have some spare time to try to get it working with my kona

**Comment by cdnninja** [comment]
No real docs this is all down via trial and error with sniffing the traffic.  I would start with locally trying the v2 endpoint for this and see if it works for you.    Switching this line: https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api/blob/444b07ca6f1820ca30989fdef2e53836e5226976/hyundai_kia_connect_api/KiaUvoApiEU.py#L1025C20-L1025C31 

to having the v2 endpoint referenced further up in the file.     Ideally work locally with the python API to start without home assistant.

**Comment by Bubo08** [comment]
My pull request might solve your problem https://github.com/Hyundai-Kia-Connect/hyundai_kia_connect_api/pull/480. 

If you have some spare time and the know-how you might try it out. Your testing help would be appreciated :-)

**Comment by cdnninja** [comment]
I have merged Bubo08s changes.  Please try the latest.

**Comment by jwillemsen** [comment]
Fixed with 2.19.0
